Replace optimist with nconf and look for .httpserverrc settings file

timothykang opened the pull request:
I thought JSHint's method of searching up the directory tree would work nicely, as it would allow projects to share settings while also allowing overrides.

Related to #67, closes #24.

indexzero commented:
Cool idea! Let me review this weekend.

indexzero commented on bin/http-server:
```diff
@@ -4,11 +4,16 @@ var colors = require('colors'),
-      .argv;
+    fs = require('fs'),
+    path = require('path'),
+    nconf = require('nconf')
```
Some people use this as a library. We should really be doing:

``` js
var nconf = require('nconf');

var config = new nconf.Provider()
  .argv({
    d: { boolean: true, default: true },
    i: { boolean: true, default: true },
    cors: { boolean: true }
  });
```

indexzero commented on package.json:
```diff
@@ -59,9 +59,9 @@
-    "optimist": "0.6.x",
     "union": "~0.4.3",
     "ecstatic": "~0.6.0",
+    "nconf": "^0.7.1",
```
Just released `nconf@0.8.x` this past weekend. Lets upgrade to that.

indexzero commented:
+1 if you can rebase of `master`

timothykang commented:
Rebased and addressed comments. In the process I undid a small change I made to how the boolean `showDir` and `autoIndex` options were being parsed that would've broken backwards compatibility.

indexzero commented:
Will take a look at this over the weekend or next week.

timothykang commented:
Rebased onto latest `master`.

timothykang commented:
Noticed the recent release, rebased onto `master` again.

kilianc commented:
+1

timothykang commented:
An oldie indeed. Rebased onto current master, in light of the renewed interest.

thornjad commented:
@timothykang I worked on replacing optimist with minimist in #464, which I still believe is a step in the right direction, but I have to say I like this nconf approach better! 

Would you mind taking a look at the merge confict in bin/http-server? Could you also look into specifying the current version of nconf, and make sure it doesn't break anything?

Thank you!

timothykang commented:
@thornjad Done!

thornjad commented on bin/http-server:
```diff
@@ -45,14 +46,32 @@ if (argv.h || argv.help) {
+
+var port = config.get('p') || config.get('port') || parseInt(process.env.PORT, 10),
+    host = config.get('a') || '0.0.0.0',
+    ssl = !!config.get('S') || !!config.get('ssl'),
```
I _think_ we can use `parseValues: true` in `.argv` above rather than need to coerce here. Correct me if I'm wrong on that, I haven't used nconf before.

timothykang commented:
@thornjad Rebased. I added value parsing for numbers and where it made sense, but I left the keys as the short forms since not every option had a long form (they could be easily added in a follow-on PR).

karlhorky commented:
I guess this can probably be closed, now that #622 was merged instead (switch `optimist` to `minimist`)
